src/AWS/Lambda/Events/ApiGateway/ProxyRequest.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE FlexibleInstances     #-}
+
+{-|
+Module      : AWS.Lambda.Events.ApiGateway.ProxyRequest
+Description : Data types that represent typical lambda responses
+Copyright   : (c) Nike, Inc., 2018
+License     : BSD3
+Maintainer  : nathan.fairhurst@nike.com, fernando.freire@nike.com
+Stability   : stable
+-}
+module AWS.Lambda.Events.ApiGateway.ProxyRequest
+    ( ProxyRequest(..)
+    , RequestContext(..)
+    , Identity(..)
+    , NoAuthorizer
+    , StrictlyNoAuthorizer
+    ) where
+
+import           Control.Monad               (mzero)
+import           Data.Aeson                  (FromJSON, Value (Object),
+                                              parseJSON, (.!=), (.:), (.:?))
+import           Data.ByteString.Base64.Lazy (decodeLenient)
+import           Data.ByteString.Lazy        (ByteString)
+import           Data.CaseInsensitive        (CI, mk)
+import           Data.HashMap.Strict         (HashMap, foldrWithKey, insert)
+import           Data.Text                   (Text)
+import qualified Data.Text.Lazy              as TL
+import qualified Data.Text.Lazy.Encoding     as TLE
+import           GHC.Generics                (Generic (..))
+
+data Identity = Identity
+    { cognitoIdentityPoolId         :: Maybe Text
+    , accountId                     :: Maybe Text
+    , cognitoIdentityId             :: Maybe Text
+    , caller                        :: Maybe Text
+    , apiKey                        :: Maybe Text
+    , sourceIp                      :: Text
+    , accessKey                     :: Maybe Text
+    , cognitoAuthenticationType     :: Maybe Text
+    , cognitoAuthenticationProvider :: Maybe Text
+    , userArn                       :: Maybe Text
+    , apiKeyId                      :: Maybe Text
+    , userAgent                     :: Maybe Text
+    , user                          :: Maybe Text
+    } deriving (Generic)
+
+instance FromJSON Identity
+
+data RequestContext a = RequestContext
+    { path              :: Text
+    , accountId         :: Text
+    , authorizer        :: Maybe a
+    , resourceId        :: Text
+    , stage             :: Text
+    , domainPrefix      :: Maybe Text
+    , requestId         :: Text
+    , identity          :: Identity
+    , domainName        :: Maybe Text
+    , resourcePath      :: Text
+    , httpMethod        :: Text
+    , extendedRequestId :: Maybe Text
+    , apiId             :: Text
+    }
+
+instance FromJSON a => FromJSON (RequestContext a) where
+    parseJSON (Object v) =
+        RequestContext <$> v .: "path" <*> v .: "accountId" <*>
+        v .:? "authorizer" <*>
+        v .: "resourceId" <*>
+        v .: "stage" <*>
+        v .:? "domainPrefix" <*>
+        v .: "requestId" <*>
+        v .: "identity" <*>
+        v .:? "domainName" <*>
+        v .: "resourcePath" <*>
+        v .: "httpMethod" <*>
+        v .:? "extendedRequestId" <*>
+        v .: "apiId"
+    parseJSON _ = mzero
+
+-- TODO: Should also include websocket fields
```

Comment:
Created a card here: https://github.com/Nike-Inc/hal/issues/53

Other than adding the fields, is there anything else we need to support websockets from apig?

Reply:
So this is a bit of an odd one.  The more I think about it, the more I think that this should be a separate event that maybe has some common models (Identity ab RequestContext maybe).  But ultimately it makes a bunch of fields optional if we try to do them together.  Which makes sense for JSON, but less so here—where we want to strongly type and communicate guarantees wherever possible.

The downside of the separate is a bunch of boilerplate.  The record field destructuring is super useful for these types, but it doesn’t seem to compose for reuse very nicely.

Reply:
I'll have to chew on this one a bit more before responding more completely